### ISSUE TITLE
Tensorflow wasm module with python bindings

### ISSUE THREAD

**Issue by jgmize** [pyodide/pyodide#50]

**Comment by jgmize** [comment]
https://github.com/tomasreimers/tensorflow-emscripten

**Comment by rth** [comment]
There is a WebAssembly backend to TensorFlow.js https://github.com/tensorflow/tfjs/tree/master/tfjs-backend-wasm

Given the complexity of Tensorflow I don't think we would be able to maintain something like it in pyodide. Closing. Please comment if you disagree.

**Comment by Shahin-rmz** [comment]
disagreed.
The reason python is python, is it's popularity among non developers and scientists.
Javascript has absolutely no place there.

**Comment by rth** [comment]
Building Tensorflow (or possibly Tensorflow Lite) with emscripten and pyodide would certainly be useful. I'm just saying that it would likely require significant engineering resources which we don't have at the moment. You are very welcome to investigate it though.

**Comment by sylvain-ri** [comment]
Having either tensorflor or pytorch #1625 support in Pyodide would give access to Deep Learning to the community. When resources allow for its development, I think it will be a great asset.

In the meantime, maybe a tutorial and an API to tfjs would be a compromise ?

**Comment by TheRook** [comment]
This is extremely useful seeing as the backend costs for tenserflow are both difficult to scale and also out of reach for non-profit projects.

**Comment by Shahin-rmz** [comment]
> This is extremely useful seeing as the backend costs for tenserflow are both difficult to scale and also out of reach for non-profit projects.

Pytorch would be also the same?

**Comment by TheRook** [comment]
> > This is extremely useful seeing as the backend costs for tenserflow are both difficult to scale and also out of reach for non-profit projects.
> 
> Pytorch would be also the same?

Yes. Pytorch is also very expensive and widely used in generative art... poor artists also need neural transfers!  Having a more general python-distuitils for Wasm would be ideal, do these bindings need to be manually built or can we use reflection to automatically generate bindings?

**Comment by lucaswiman** [comment]
Just a few notes I gathered on this:
* For Pytorch, you can add an extension as discussed in https://pytorch.org/tutorials/advanced/extend_dispatcher.html This would involve implementing a `TensorImpl` backed by WebGPU as well as implementing the operations described in the docs.
* Tensorflow has the notion of a "pluggable device" framework (https://www.tensorflow.org/install/gpu_plugins and https://github.com/tensorflow/community/blob/master/rfcs/20200624-pluggable-device-for-tensorflow.md) that supports several backends already.

**Comment by hpssjellis** [comment]
@rth 

What would be really useful would be the ability to convert keras hd5  files into tFlite files using the TFLiteConverter.  See the example code below.

```
import tensorflow as tf
# Convert the model
converter = tf.lite.TFLiteConverter.from_keras_model_file('model.h5') # path to the HD5 file
tflite_model = converter.convert()
# Save the model.
with open('model.tflite', 'wb') as f:
  f.write(tflite_model)


```
OR
```

import tensorflow as tf

# Load the Keras model
keras_model = tf.keras.models.load_model('your_keras_model.h5')

# Convert the Keras model to a TensorFlow Lite model
converter = tf.lite.TFLiteConverter.from_keras_model(keras_model)
tflite_model = converter.convert()

# Save the TensorFlow Lite model to a file
with open('converted_model.tflite', 'wb') as f:
    f.write(tflite_model)

```


TensorflowJS does most of the web abilities, but converting tensorflowJS files to tFlite files is a gap in their support. I believe Tensorflowjs can export Keras hd5 files, but presently converting Keras to TFlite for use on microcontrollers is only supported with Python.


Even if it could only convert fully dense connected layers that would be better than loading an entire python development environment for this one conversion.  Any opinions ?
